tools/stats/monitor.py

```diff
@@ -11,136 +24,314 @@
-    processes = get_processes_running_python_tests()
-    per_process_info = []
-    for p in processes:
-        info = {
-            "pid": p.pid,
-            "cmd": " ".join(p.cmdline()),
-            "cpu_percent": p.cpu_percent(),
-            "rss_memory": p.memory_info().rss,
+try:
+    import pynvml  # type: ignore[import]
+
+    _HAS_PYNVML = True
+except ModuleNotFoundError:
+    pass
+
+try:
+    import amdsmi  # type: ignore[import]
+
+    _HAS_AMDSMI = True
+except ModuleNotFoundError:
+    pass
+
+
+def parse_args() -> argparse.Namespace:
+    """
+    Parse command line arguments.
+
+    Returns:
+        argparse.Namespace: Parsed arguments.
+    """
+    parser = argparse.ArgumentParser(description=" System-level Usage Logger ")
+
+    # debug mode used in local to gracefully exit the script when ctrl+c is pressed, and print out the json output in a pretty format.
+    parser.add_argument("--debug", action="store_true", help="Enable debug mode")
+
+    parser.add_argument(
+        "--log-interval",
+        type=int,
+        default=5,
+        help="set time interval for logging utilization data, default is 5 seconds",
+    )
+    args = parser.parse_args()
+    return args
+
+
+class UsageLogger:
+    """
+    Collect and display usage data, including CPU, memory, GPU memory utilization, and GPU utilization.
+    """
+
+    def __init__(
+        self,
+        log_interval=5,
+        is_debug_mode=False,
+        pynvml_enabled=False,
+        amdsmi_enabled=False,
+    ) -> None:
+        """
+        log_interval: Time interval in seconds for collecting usage data; default is 5 seconds.
+        is_debug_mode: Useful if you're testing on a local machine and want to see the output in a pretty format with more information.
+        """
+        self._log_interval = log_interval
+        self._summary_info = {
+            "level": "metadata",
+            "interval": self._log_interval,
         }
+        self._has_pynvml = pynvml_enabled
+        self._has_amdsmi = amdsmi_enabled
+        self._kill_now = False
+        self._gpu_handles = []
+        self._gpu_libs_detected = []
+        self._num_of_cpus = 0
+        self._debug_mode = is_debug_mode
+        self._initial_gpu_handler()
 
-        # https://psutil.readthedocs.io/en/latest/index.html?highlight=memory_full_info
-        # requires higher user privileges and could throw AccessDenied error, i.e. mac
-        try:
-            memory_full_info = p.memory_full_info()
+    def start(self) -> None:
+        """
+        runs the main loop of the program.
+        the firstt json record is the metadata of the run, including the start time, end time, and the interval of the log.
+        """
 
-            info["uss_memory"] = memory_full_info.uss
-            if "pss" in memory_full_info:
-                # only availiable in linux
-                info["pss_memory"] = memory_full_info.pss
+        self._summary_info["start_time"] = datetime.datetime.now(
```

Comment:
Do you want to have an entry for `stop_time` at the end just to pair with this one?

Reply:
I will add that in next prs,  want to test and add interraptable sleep. otherwise, when we gracefully exit after killing pid, the while loop might still stuck in sleep, this may cause some write issues